File: Codex/Python/2025/Day1b_Solution_2025.py
def count_zero_hits(position: int, direction: str, distance: int) -> int:
    if direction == "L":
        first_hit = position if position != 0 else 100
    else:
        first_hit = (100 - position) if position != 0 else 100

    if distance < first_hit:
        return 0

    return 1 + (distance - first_hit) // 100


def main() -> None:
    position = 50
    zero_count = 0

    with open("Day1_Input.txt", "r", encoding="utf-8") as file:
        for raw_line in file:
            line = raw_line.strip()
            if not line:
                continue

            direction = line[0]
            distance = int(line[1:])

            zero_count += count_zero_hits(position, direction, distance)

            if direction == "L":
                position = (position - distance) % 100
            else:
                position = (position + distance) % 100

    print(zero_count)


if __name__ == "__main__":
    main()

File: Codex/Rust/2025/Day1b_Solution_2025.rs
use std::fs;

fn zero_hits(position: u32, direction: u8, distance: u128) -> u128 {
    let first = match direction {
        b'L' => {
            if position == 0 {
                100
            } else {
                position as u128
            }
        }
        b'R' => {
            let steps_to_zero = (100 - position) % 100;
            if steps_to_zero == 0 {
                100
            } else {
                steps_to_zero as u128
            }
        }
        _ => panic!("invalid rotation direction"),
    };

    if distance < first {
        0
    } else {
        1 + (distance - first) / 100
    }
}

fn main() {
    let input = fs::read_to_string("Day1_Input.txt").expect("failed to read Day1_Input.txt");

    let mut position: u32 = 50;
    let mut answer: u128 = 0;

    for line in input.lines().filter(|line| !line.trim().is_empty()) {
        let bytes = line.as_bytes();
        let direction = bytes[0];
        let distance: u128 = line[1..]
            .trim()
            .parse()
            .expect("invalid rotation distance");

        answer += zero_hits(position, direction, distance);

        let step = (distance % 100) as u32;
        position = match direction {
            b'L' => (position + 100 - step) % 100,
            b'R' => (position + step) % 100,
            _ => panic!("invalid rotation direction"),
        };
    }

    println!("{answer}");
}
